mcp/src/test/java/io/modelcontextprotocol/server/McpAsyncServerExchangeTests.java

```diff
@@ -0,0 +1,697 @@
+	}
+
+	@Test
+	void testGetClientInfo() {
+		assertThat(exchange.getClientInfo()).isEqualTo(clientInfo);
+	}
+
+	// ---------------------------------------
+	// Logging Notification Tests
+	// ---------------------------------------
+
+	@Test
+	void testLoggingNotificationWithNullMessage() {
+		StepVerifier.create(exchange.loggingNotification(null)).verifyErrorSatisfies(error -> {
+			assertThat(error).isInstanceOf(McpError.class).hasMessage("Logging message must not be null");
+		});
+	}
+
+	@Test
+	void testLoggingNotificationWithAllowedLevel() {
+
+		McpSchema.LoggingMessageNotification notification = McpSchema.LoggingMessageNotification.builder()
+			.level(McpSchema.LoggingLevel.ERROR)
+			.logger("test-logger")
+			.data("Test error message")
+			.build();
+
+		when(mockSession.sendNotification(eq(McpSchema.METHOD_NOTIFICATION_MESSAGE), eq(notification)))
+			.thenReturn(Mono.empty());
+
+		StepVerifier.create(exchange.loggingNotification(notification)).verifyComplete();
+	}
+
+	@Test
+	void testLoggingNotificationWithFilteredLevel() {
+		// Given - Set minimum level to WARNING, send DEBUG message
+		exchange.setMinLoggingLevel(McpSchema.LoggingLevel.WARNING);
+
+		McpSchema.LoggingMessageNotification debugNotification = McpSchema.LoggingMessageNotification.builder()
+			.level(McpSchema.LoggingLevel.DEBUG)
+			.logger("test-logger")
+			.data("Debug message that should be filtered")
+			.build();
+
+		// When & Then - Should complete without sending notification
+		StepVerifier.create(exchange.loggingNotification(debugNotification)).verifyComplete();
+	}
+
+	@Test
+	void testLoggingNotificationLevelFiltering() {
+		// Given - Set minimum level to WARNING
+		exchange.setMinLoggingLevel(McpSchema.LoggingLevel.WARNING);
+
+		// Test DEBUG (should be filtered)
+		McpSchema.LoggingMessageNotification debugNotification = McpSchema.LoggingMessageNotification.builder()
+			.level(McpSchema.LoggingLevel.DEBUG)
+			.logger("test-logger")
+			.data("Debug message")
+			.build();
+
+		StepVerifier.create(exchange.loggingNotification(debugNotification)).verifyComplete();
+
+		// Test INFO (should be filtered)
+		McpSchema.LoggingMessageNotification infoNotification = McpSchema.LoggingMessageNotification.builder()
+			.level(McpSchema.LoggingLevel.INFO)
+			.logger("test-logger")
+			.data("Info message")
+			.build();
+
+		StepVerifier.create(exchange.loggingNotification(infoNotification)).verifyComplete();
+
+		// Test WARNING (should be sent)
+		McpSchema.LoggingMessageNotification warningNotification = McpSchema.LoggingMessageNotification.builder()
+			.level(McpSchema.LoggingLevel.WARNING)
+			.logger("test-logger")
+			.data("Warning message")
+			.build();
+
+		when(mockSession.sendNotification(eq(McpSchema.METHOD_NOTIFICATION_MESSAGE), eq(warningNotification)))
+			.thenReturn(Mono.empty());
+
+		StepVerifier.create(exchange.loggingNotification(warningNotification)).verifyComplete();
```

Comment:
It looks like the interaction that sends the notification and one that filters it out do complete the same way - how do we distinguish them? Can we ensure the mock fails if it receives an interaction it's not supposed to get?